src/_zkapauthorizer/server/spending.py

```diff
@@ -0,0 +1,69 @@
+from __future__ import absolute_import, division, print_function, unicode_literals
+
+try:
+    from typing import Any
+except ImportError:
+    pass
+
+import attr
+from challenge_bypass_ristretto import PublicKey
+from prometheus_client import CollectorRegistry
+from twisted.internet.interfaces import IReactorTime
+from zope.interface import Interface, implementer
+
+
+class ISpender(Interface):
+    """
+    An ``ISpender`` can records spent ZKAPs and reports double spends.
+    """
+
+    def mark_as_spent(public_key, passes):
+        # type: (PublicKey, list[bytes]) -> None
+        """
+        Record the given ZKAPs (associated to the given public key as having
+        been spent.
+
+        This does *not* report errors and should only be used in cases when
+        recording spending that has already happened. This can be because
+        we could not contact the spending service when they were spent, or
+        because we can't yet check before making changes to the node.
+        """
```

Comment:
This doesn't seem to match the usage in the implementation so I'm not sure I understand the direction of the PR overall.

Is the docstring right?  If so, then isn't this more like an API for storage nodes to invoke when the spending service comes back online after a period of unavailability with all of the passes they observed during the unavailable time?

Or is the implementation doing the right thing in calling this, in which case errors (like double spend attempts) should come out of this and stop the storage operation?

Or ... something else?

Reply:
> Is the docstring right? If so, then isn't this more like an API for storage nodes to invoke when the spending service comes back online after a period of unavailability with all of the passes they observed during the unavailable time?

Yes, the end state should be this only gets called in those cases.

> Or is the implementation doing the right thing in calling this, in which case errors (like double spend attempts) should come out of this and stop the storage operation?

Some operations, like changes to mutable files, can't be rolled back if we get a failure, so we need a two phase process to handle errors. If we call *before* the operation and record as spent, we potentially charge users for an operation that fails; if we call *after* the operation and report errors, we return an error for a successful operation which will cause clients to be confused.

Given that we *do* need an API like is described here for recovering from unavailability, I chose to implement just that API as a first pass, which allows us to *detect* but not *prevent* double spending. We will also need a pair of APIs that checks for double spends and reports errors, followed by committing to spending those passes.